packages/next/types/index.d.ts

```diff
@@ -65,17 +65,30 @@ export {
   NextApiHandler,
 }
 
-export type GetStaticProps<
-  P extends { [key: string]: any } = { [key: string]: any },
-  Q extends ParsedUrlQuery = ParsedUrlQuery
-> = (ctx: {
+export type GetStaticPropsContext<Q extends ParsedUrlQuery = ParsedUrlQuery> = {
   params?: Q
   preview?: boolean
   previewData?: any
-}) => Promise<{
+}
+
+export type GetStaticPropsResult<P> = {
   props: P
   unstable_revalidate?: number | boolean
-}>
+}
+
+export type GetStaticProps<
+  P extends { [key: string]: any } = { [key: string]: any },
+  Q extends ParsedUrlQuery = ParsedUrlQuery
+> = (context: GetStaticPropsContext<Q>) => Promise<GetStaticPropsResult<P>>
+
+export type StaticProps<
+  T,
+  Q extends ParsedUrlQuery = ParsedUrlQuery
+> = T extends (
+  context?: GetStaticPropsContext<Q>
+) => Promise<GetStaticPropsResult<infer P>>
+  ? GetStaticPropsResult<P>['props']
```

Comment:
What's the difference between this line and `? P` ?

Reply:
In my understanding `P` are the `Props` passed to `GetStaticProps<Props>` that have to be manually declared upfront, right?

`GetStaticPropsResult['props']` would be the inferred types after the Promise was resolved.

But as I said I am no typescript expert and could easily be wrong about this.

Reply:
What I mean is, can we make it work like this:
```ts
) => Promise<GetStaticPropsResult<infer P>>
  ? P
```
`infer P` should make `P` available, so there's no need to go for `props` inside the type.